### ISSUE TITLE
User without access to default text format can't edit field

### ISSUE THREAD

**Issue by BWPanda** [backdrop/backdrop-issues#4076]
**Description of the bug**
If you have a user who doesn't have access to, say, the Full HTML format, and if you set the [default text format](https://github.com/backdrop/backdrop-issues/issues/3378) for a field to Full HTML, then that user cannot edit that field on new node forms.

**Steps To Reproduce**
To reproduce the behavior:
1. Create a new user with just the Authenticated role.
2. Make sure the Full HTML format cannot be used by Authenticated users (disable that permission).
3. Edit the Body field of the Page content type and set the default text format to Full HTML.
4. As the new, Authenticated user, create a new Page node.

**Actual behavior**
Body field is disabled, saying "This field has been disabled because you do not have sufficient permissions to edit it."

**Expected behavior**
Body field is editable, with the available text formats being ones the user has access to (not Full HTML).

**Additional information**
One might argue that this is expected functionality, as setting Full HTML means you only want users to use that. And if they don't have access they shouldn't be able to use that field. But to that I say that you're setting the *default* text format, not necessarily the one that *has* to be used. For that see https://github.com/backdrop/backdrop-issues/issues/2615.

**Comment by BWPanda** [comment]
Here's a [PR](https://github.com/backdrop/backdrop/pull/2893) that fixes this by not hiding/disabling the field when the user doesn't have access to the default text format and the field is empty. In that case, just show the text formats the user does have access to.

**Comment by klonos** [comment]
> *Expected behavior*
>Body field is editable, with the available text formats being ones the user has access to (not Full HTML).

Hmm, 🤔 ...I have double-checked on a vanilla D7 site, just to be sure, and this is not the default behavior. Note that in order to test this, I had to actually add a dot in the field default value, before changing the format and finally saving the field settings - that's because of https://www.drupal.org/project/drupal/issues/784672, which is still a not solved in D7 (We have fixed this in #3378).

I guess it's down to what the expectation/interpretation of the "Default value" field setting is. Consider the use case where the admin has set the default value to be in full HTML, whereas the content editor does not have access to that format. When the editor tries to create a new piece of content with this fix here, then they will not get any default value in that field (but still be able to create content - using the formats they are allowed to use). Is this expected behavior, or not? 🤷‍♂

One can take it as that the default value is merely a recommendation/placeholder, so does not really matter (the editor can remove that default value completely anyway). Others may actually want to mandate something specific, in which case, it won't work the way they expect. Perhaps the way to go is to provide a way to "lock" the format (#2615).

I am not opposed to this change here; just need to wrap my head around how things work, so that we don't break existing functionality.

**Comment by BWPanda** [comment]
With any other field in Backdrop, setting a default value does not restrict authors from changing it. Without this patch, we're restricting authors to a certain format in certain situations. That provides a bad and inconsistent UX in my opinion.

As we've already linked to, the other issue re. allowed formats provides for the situation where you *do* want to restrict authors to one or more formats, so this issue should just be focused on setting an optional default.

**Comment by klonos** [comment]
> ...so this issue should just be focused on setting an optional default.

Agreed 👍 ...thanks for adding the "needs feedback" tag; I'd like to have more eyes on this.

**Comment by olafgrabienski** [comment]
Strange, when #3378 was merged, @quicksketch wrote in https://github.com/backdrop/backdrop-issues/issues/3378#issuecomment-481881736:

> I was pleased to see that if a user doesn't have access to a format on /node/add/post, the text format just selects the first format they do have access to as long as the text string is empty.

So, has the current issue been introduced afterwards?

**Comment by olafgrabienski** [comment]
>> ...so this issue should just be focused on setting an optional default.

> Agreed 👍 ...thanks for adding the "needs feedback" tag; I'd like to have more eyes on this.

I also agree that the default shouldn't be a restriction.

**Comment by BWPanda** [comment]
@olafgrabienski Yes, I was confused about @quicksketch's comment too. Not sure what happened there/since...

**Comment by BWPanda** [comment]
PR has been updated with latest core changes and Tugboat sandbox.

**Comment by olafgrabienski** [comment]
Tested on the sandbox website, works for me: As user without the Raw HTML permission, I was able to add a Page with the Body field set to the default format Raw HTML. As expected, the text format fell back to Filtered HTML.

**Comment by quicksketch** [comment]
Seems like an area where we need test coverage if we've previously had this functionality and accidentally lost it at some point. We have similar testing in `testFormatWidgetPermissions` in `filter.test`, we should add the new "use default if empty" behavior check there.

**Comment by yorkshire-pudding** [comment]
Is this issue a duplicate of #5151 ?

**Comment by olafgrabienski** [comment]
> Is this issue a duplicate of https://github.com/backdrop/backdrop-issues/issues/5151 ?

Or the other way round! However interesting:

- Both issues have a PR but are missing `automated tests`.
- The PRs look quite different.

**Comment by kiamlaluno** [comment]
I would rather close the other issue as duplicate.
